### ISSUE TITLE
Intermittent `UnrecognizedScopeTarget` under JRuby

### ISSUE THREAD

**Issue by alaz** [palkan/action_policy#75]
### Environment

**Ruby Version:** JRuby 9.2.7
**Framework Version (Rails, whatever):** Rails  5.2.3
**Action Policy Version:** 0.3.1

### What did you do?

I am observing `ActionPolicy::UnrecognizedScopeTarget` occasionally upon the application's start. It happens for different models. A typical stacktrace is

```
vendor/bundle/jruby/2.5.0/gems/action_policy-0.3.1/lib/action_policy/policy/scoping.rb:105:in `resolve_scope_type': Couldn't infer scope type for Topic::ActiveRecord_Relation instance (ActionPolicy::UnrecognizedScopeTarget)
    from vendor/bundle/jruby/2.5.0/gems/action_policy-0.3.1/lib/action_policy/behaviours/scoping.rb:28:in `authorization_scope_type_for'
    from vendor/bundle/jruby/2.5.0/gems/action_policy-0.3.1/lib/action_policy/behaviours/scoping.rb:17:in `authorized_scope'
    from app/controllers/topics_suggest_controller.rb:44:in `topic_load'
```

Where `topic_load` is

```
  def topic_load
    @topic = authorized_scope(Topic.all).find params[:id]
    authorize! @topic
  end
```

This never happens during tests or development, being it MRI or JRuby. The application production always runs under JRuby.

### Experiments

Since JRuby has fair threads, I thought it could be caused by some lack of thread safety. I have seen a similar problem before https://github.com/rails/rails/issues/23699#issuecomment-379312782 , and I already have a special Rails route to pre-warm the application after start up. So I have simply put the following into that controller action:

```
    # warm up Action Policy scopes
    authorized_scope(Post.all)
```

The errors are gone. This makes me think that a problem with tread-safety in resolving scopes is very likely.

Regards,
Alexander

**Comment by palkan** [comment]
Thanks for the report!

Could you please try the version from master? I found a couple of thread-safety problems in scoping code, should be fixed now.

**Comment by alaz** [comment]
As far as I can see, it is fixed. I have re-deployed few times at 3 servers and the problem did not show up.

Thank you for the quick fix!

**Comment by palkan** [comment]
Released in 0.3.2.
